File: C/ex4/unroll.c
int unroll(int M1,int N1,int M2,int N2,double theta1[M1][N1],double theta2[M2][N2],double nn_params[M1*N1+M2*N2])
{

	int i, j, k;

	k = 0;
	for(i=0; i<M1; i++){
		for(j=0; j<N1; j++){
			nn_params[k] = theta1[i][j];
			k = k + 1;
		}
	}

	for(i=0; i<M2; i++){
		for(j=0; j<N2; j++){
			nn_params[k] = theta2[i][j];
			k = k + 1;
		}
	}	
	
	return 0;
}

int reshape(int M1,int N1,int M2,int N2,double theta1[M1][N1],double theta2[M2][N2],double nn_params[M1*N1+M2*N2])
{

	int i, j, k;

	k = 0;
	for(i=0; i<M1; i++){
		for(j=0; j<N1; j++){
			 theta1[i][j] = nn_params[k];
	         k = k +1;
		}
	}

	for(i=0; i<M2; i++){
		for(j=0; j<N2; j++){
			theta2[i][j] = nn_params[k]; 
			k = k + 1;		
		}
	}
	

	return 0;
}

File: Fortran/ex4/unroll.f90
	subroutine unroll(M1,N1,M2,N2,theta1,theta2,nn_params)
	integer:: i, j, k
	integer:: M1, N1, M2, N2
	double precision, dimension(M1,N1)::theta1
	double precision, dimension(M2,N2)::theta2
	double precision, dimension( M1*N1 + M2*N2)::nn_params

	k = 1
	do i=1,M1
	do j=1,N1 
		nn_params(k) = theta1(i,j)
		k = k + 1
	enddo
	enddo

	do i=1,M2
	do j=1,N2
		nn_params(k) = theta2(i,j)
		k = k + 1
	enddo
	enddo	

	end subroutine unroll
	
	
	subroutine reshape(M1,N1,M2,N2,theta1,theta2,nn_params)
	integer:: i, j, k
	integer:: M1, N1, M2, N2
	double precision, dimension(M1,N1)::theta1
	double precision, dimension(M2,N2)::theta2
	double precision, dimension( M1*N1 + M2*N2)::nn_params

  
	k = 1
	do i=1,M1
	do j=1,N1 
		theta1(i,j) = nn_params(k) 
		k = k + 1
	enddo
	enddo

	do i=1,M2
	do j=1,N2
	    theta2(i,j) = nn_params(k) 
		k = k + 1
	enddo
	enddo
	
	end subroutine reshape	
	


	
